modules/andes-core/broker/src/main/java/org/wso2/andes/kernel/dtx/DtxRegistry.java

```diff
@@ -0,0 +1,246 @@
+/*
+ * Copyright (c) 2016, WSO2 Inc. (http://wso2.com) All Rights Reserved.
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.wso2.andes.kernel.dtx;
+
+import org.wso2.andes.kernel.AndesAckData;
+import org.wso2.andes.kernel.AndesChannel;
+import org.wso2.andes.kernel.AndesException;
+import org.wso2.andes.kernel.AndesMessage;
+import org.wso2.andes.kernel.DtxStore;
+import org.wso2.andes.kernel.MessagingEngine;
+import org.wso2.andes.server.txn.IncorrectDtxStateException;
+import org.wso2.andes.server.txn.RollbackOnlyDtxException;
+import org.wso2.andes.server.txn.TimeoutDtxException;
+
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import javax.transaction.xa.Xid;
+
+public class DtxRegistry {
+
+    private final Map<ComparableXid, DtxBranch> branches;
+
+    private final DtxStore dtxStore;
+
+    private final MessagingEngine messagingEngine;
+
+    public DtxRegistry(DtxStore dtxStore, MessagingEngine messagingEngine) {
+        this.dtxStore = dtxStore;
+        branches = new HashMap<>();
+        this.messagingEngine = messagingEngine;
+    }
+
+    public synchronized DtxBranch getBranch(Xid xid) {
+        return branches.get(new ComparableXid(xid));
+    }
+
+    public synchronized boolean registerBranch(DtxBranch branch) {
+        ComparableXid xid = new ComparableXid(branch.getXid());
+        if(!branches.containsKey(xid)) {
+            branches.put(xid, branch);
+            return true;
+        }
+        return false;
+    }
+
+    public synchronized void prepare(Xid xid)
```

Comment:
Generally, multiple threads won't call prepare simultaneously. But we need to make sure the branch states are modified atomically.